### ISSUE TITLE
has_edge error

### ISSUE THREAD

**Issue by mbesancon** [JuliaGraphs/LightGraphs.jl#939]
In `src/linalg/spectral.jl`, `has_edge` is called as `has_edge(g,i,j)` while the interface definition is
`has_edge(g,e)`. I guess this part of linalg hadn't been updated yet.

**Comment by sbromberger** [comment]
Both are valid; that latter is an alternate convenience method. Is it causing s problem?

**Comment by mbesancon** [comment]
yes, on a custom graph type, if only defining the interface one `has_edge(g,e)`, I got an error when calling `gplot`.
The graph type uses SimpleEdge as an edge type

**Comment by sbromberger** [comment]
I’d need to check but I think that the `(g, i, j)` variant is specified such it should work with custom edge types that define `Edge(i, j)` as a constructor.

**Comment by sbromberger** [comment]
Ah, so I see the issue. `has_edge(g, i, j)` is defined for AbstractSimpleEdges; it just so happens that
1) other graph types have implemented it as well, and
2) this convenience method is used throughout LightGraphs.

So I would suggest that we make this particular method part of the formal API.

**Comment by mbesancon** [comment]
It could be nice to keep the interface minimal and to only define this one as optional. The problem is that we cannot implement `has_edge(g, i, j)` from `has_edge(g,e)`, but the other way around yes

**Comment by sbromberger** [comment]
if `g,i,j` is optional, then we'd need to change a whole bunch of functions in LG. I'm not in favor of this.

**Comment by mbesancon** [comment]
if we implement the default `has_edge(g,e)` as above, we can only require `has_edge(g,i,j)`

**Comment by sbromberger** [comment]
`has_edge(g,e) = has_edge(g,src(e),dst(e))` is the most logical way of doing this, but one of the problems I had was with edges that hold metadata. Does an edge from `i` to `j` exist if the weights are different? Consider:

```
g = SimpleWeightedGraph(PathGraph(3))
e = SimpleWeightedEdge(1, 2, 5.0)
```
what should `has_edge(g, e)` return? (Right now, it's `true`, but this has bothered me for a while.)

**Comment by mbesancon** [comment]
in that case, it should remain so, weights are floats anyway you would run into trouble if comparing directly `w == w0`.
That's why `has_edge(g,e) = has_edge(g,src(e),dst(e))` would be the default implementation but could be overwritten by new graph types

**Comment by mbesancon** [comment]
These don't need to change don't they? In fact having the two has_edge would solve the problem of graphs where you can have "two definitions" of equality (meta, weighted):
- one checks whether there is any edge between i and j
- the other checks if there is a specific edge e

**Comment by mbesancon** [comment]
`add_edge!` and `rem_edge!` can remain as such, otherwise we won't be able to have a unified signature `f(g,e)` for these

**Comment by sbromberger** [comment]
> * one checks whether there is any edge between i and j
> * the other checks if there is a specific edge e

This makes sense and is consistent, but would be a major breaking change. The time is right, though.

**Comment by mbesancon** [comment]
I'm not sure if it is technically a breaking change since has_edge(g,i,j) was already required for some functionalities, so the "breaking" was already there. 
But yes in anyway this is a good thing to do before 1.0

**Comment by sbromberger** [comment]
It's breaking in that `has_edge(g, e)` will now be required to check ALL edge attributes for equality (that is, for SWG, it will check the weights, and for MetaGraphs, it will check `eprops`). It's not currently doing so.

**Comment by matbesancon** [comment]
agreed. It's a change in the definition but changing the interface will not break graph implementations immediately.

**Comment by sbromberger** [comment]
It very well might, as we are changing the return value of `has_edge` for edges with different weights (from true to false). That is, the tests for SWG and MetaGraphs will certainly fail.

**Comment by matbesancon** [comment]
yes, so it will require a new version and breaking change warning in these packages precisely

**Comment by jpfairbanks** [comment]
I agree that we should require `has_edge(g, i,j)` and have a default override `has_edge(g,e::Edge) = has_edge(g, src(e), dst(e))`, so that if a graph type implementer wants to use `has_edge(g,e)` to mean "this specific edge with these properties" then they override the default behavior.

**Comment by sbromberger** [comment]
Yes, I think we're all in agreement here. The API will change from `has_edge(g, e)` to `has_edge(g, i, j)`, and it will only check to see whether src and dst are in the graph. For all graph types, we will encourage the creation of `has_edge(g,e)` which will check the literal equality of the edges in the graph and will return `true` *iff* `e` is `==` some other edge in the graph. (Edge equality will be left  to the individual graph type.)
